File: mobile/ios/Core/Network/APIClient.swift
import Foundation

struct APIClient {
    enum APIError: Error {
        case badURL
        case http(Int, String)
        case decode(String)
    }

    let baseURL: String
    let token: String

    private static let encoderISO: ISO8601DateFormatter = {
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        return f
    }()

    private static let decoder: JSONDecoder = {
        let d = JSONDecoder()
        d.dateDecodingStrategy = .custom { decoder in
            let c = try decoder.singleValueContainer()
            let raw = try c.decode(String.self)
            if let ts = Fmt.isoFrac.date(from: raw) ?? Fmt.iso.date(from: raw) {
                return ts
            }
            throw DecodingError.dataCorruptedError(in: c, debugDescription: "invalid RFC3339 date: \(raw)")
        }
        return d
    }()

    private func makeURL(path: String, query: [URLQueryItem] = []) throws -> URL {
        guard var comp = URLComponents(string: baseURL) else { throw APIError.badURL }
        let safeBasePath = comp.path.hasSuffix("/") ? String(comp.path.dropLast()) : comp.path
        comp.path = safeBasePath + path
        comp.queryItems = query.isEmpty ? nil : query
        guard let url = comp.url else { throw APIError.badURL }
        return url
    }

    private func request(path: String, method: String = "GET", query: [URLQueryItem] = [], body: Data? = nil) throws -> URLRequest {
        let url = try makeURL(path: path, query: query)
        var req = URLRequest(url: url)
        req.httpMethod = method
        req.timeoutInterval = 20
        if !token.isEmpty {
            req.addValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
        if body != nil {
            req.addValue("application/json", forHTTPHeaderField: "Content-Type")
        }
        req.httpBody = body
        return req
    }

    private func send(_ req: URLRequest) async throws -> Data {
        let (data, resp) = try await URLSession.shared.data(for: req)
        let code = (resp as? HTTPURLResponse)?.statusCode ?? 500
        if code == 401 {
            await MainActor.run { AuthGate.shared.handleUnauthorized() }
            throw APIError.http(401, "unauthorized")
        }
        guard (200..<300).contains(code) else {
            throw APIError.http(code, String(data: data, encoding: .utf8) ?? "http error")
        }
        return data
    }

    private func decode<T: Decodable>(_ type: T.Type, from data: Data) throws -> T {
        do {
            return try Self.decoder.decode(type, from: data)
        } catch {
            throw APIError.decode(String(describing: error))
        }
    }

    func login(username: String, password: String) async throws -> LoginResponse {
        let body = try JSONEncoder().encode(LoginRequest(username: username, password: password))
        do {
            let data = try await send(try request(path: "/auth/mobile/login", method: "POST", body: body))
            return try decode(LoginResponse.self, from: data)
        } catch {
            let data = try await send(try request(path: "/login", method: "POST", body: body))
            return try decode(LoginResponse.self, from: data)
        }
    }

    func fetchDevices() async throws -> [Device] {
        let data = try await send(try request(path: "/devices"))
        return try decode([Device].self, from: data)
    }

    func fetchDevice(id: String) async throws -> Device {
        let data = try await send(try request(path: "/devices/\(id)"))
        return try decode(Device.self, from: data)
    }

    func search(keyword: String) async throws -> [GlobalSearchItem] {
        let q = keyword.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !q.isEmpty else { return [] }
        let data = try await send(try request(path: "/search", query: [URLQueryItem(name: "q", value: q)]))
        return try decode([GlobalSearchItem].self, from: data)
    }

    func fetchDeviceHistory(type: String, id: String, start: Date, end: Date, maxPoints: Int, interval: String) async throws -> [DeviceHistoryPoint] {
        let query = [
            URLQueryItem(name: "type", value: type),
            URLQueryItem(name: "id", value: id),
            URLQueryItem(name: "start", value: Self.encoderISO.string(from: start)),
            URLQueryItem(name: "end", value: Self.encoderISO.string(from: end)),
            URLQueryItem(name: "max_points", value: String(maxPoints)),
            URLQueryItem(name: "interval", value: interval)
        ]
        let data = try await send(try request(path: "/metrics/history", query: query))
        return try decode(HistoryResponse<DeviceHistoryPoint>.self, from: data).data
    }

    func fetchTraffic(id: String, start: Date, end: Date, maxPoints: Int, interval: String) async throws -> [TrafficHistoryPoint] {
        let query = [
            URLQueryItem(name: "type", value: "traffic"),
            URLQueryItem(name: "id", value: id),
            URLQueryItem(name: "start", value: Self.encoderISO.string(from: start)),
            URLQueryItem(name: "end", value: Self.encoderISO.string(from: end)),
            URLQueryItem(name: "max_points", value: String(maxPoints)),
            URLQueryItem(name: "interval", value: interval)
        ]
        let data = try await send(try request(path: "/metrics/history", query: query))
        return try decode(HistoryResponse<TrafficHistoryPoint>.self, from: data).data
    }
}

File: mobile/android/app/src/main/java/com/netpulse/mobile/data/ApiClient.kt
package com.netpulse.mobile.data

import kotlinx.serialization.builtins.ListSerializer
import kotlinx.serialization.builtins.serializer
import kotlinx.serialization.json.Json
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.net.URLEncoder
import java.time.Instant
import java.util.concurrent.TimeUnit

class ApiClient(private val session: SessionStore) {
    private val json = Json { ignoreUnknownKeys = true }
    private val http = OkHttpClient.Builder()
        .connectTimeout(15, TimeUnit.SECONDS)
        .readTimeout(25, TimeUnit.SECONDS)
        .build()
    private val media = "application/json; charset=utf-8".toMediaType()

    private fun url(path: String, query: Map<String, String?> = emptyMap()): String {
        val base = session.baseUrl.trimEnd('/')
        val qs = query.filterValues { !it.isNullOrBlank() }.map { (k, v) ->
            "${enc(k)}=${enc(v.orEmpty())}"
        }.joinToString("&")
        return base + path + if (qs.isBlank()) "" else "?$qs"
    }

    private fun enc(v: String) = URLEncoder.encode(v, Charsets.UTF_8.name())

    private fun request(path: String, query: Map<String, String?> = emptyMap(), method: String = "GET", body: String? = null): Request {
        val builder = Request.Builder().url(url(path, query))
        if (session.token.isNotBlank()) builder.addHeader("Authorization", "Bearer ${session.token}")
        if (method != "GET") builder.method(method, body.orEmpty().toRequestBody(media))
        return builder.build()
    }

    private fun send(req: Request): String {
        http.newCall(req).execute().use { resp ->
            val text = resp.body?.string().orEmpty()
            if (resp.code == 401) {
                session.clear()
                throw IllegalStateException("登录已过期，请重新登录")
            }
            if (!resp.isSuccessful) throw IllegalStateException(text.ifBlank { "HTTP ${resp.code}" })
            return text
        }
    }

    fun login(username: String, password: String): LoginResponse {
        val body = json.encodeToString(LoginRequest.serializer(), LoginRequest(username, password))
        val first = runCatching { send(request("/auth/mobile/login", method = "POST", body = body)) }
            .recoverCatching { send(request("/login", method = "POST", body = body)) }
            .getOrThrow()
        val resp = json.decodeFromString(LoginResponse.serializer(), first)
        session.token = resp.token
        return resp
    }

    fun fetchDevices(): List<Device> = json.decodeFromString(ListSerializer(Device.serializer()), send(request("/devices")))

    fun fetchDevice(id: Long): Device = json.decodeFromString(Device.serializer(), send(request("/devices/$id")))

    fun fetchPort(id: Long): Port = json.decodeFromString(Port.serializer(), send(request("/interfaces/$id")))

    fun search(q: String): List<SearchItem> = if (q.isBlank()) emptyList() else json.decodeFromString(
        ListSerializer(SearchItem.serializer()),
        send(request("/search", mapOf("q" to q)))
    )

    fun fetchDeviceHistory(type: String, id: Long, start: Instant, end: Instant, maxPoints: Int, interval: String): List<DeviceHistoryPoint> {
        val text = send(request("/metrics/history", mapOf(
            "type" to type,
            "id" to id.toString(),
            "start" to start.toString(),
            "end" to end.toString(),
            "max_points" to maxPoints.toString(),
            "interval" to interval
        )))
        return json.decodeFromString(HistoryResponse.serializer(DeviceHistoryPoint.serializer()), text).data
    }

    fun fetchTraffic(id: Long, start: Instant, end: Instant, maxPoints: Int, interval: String): List<TrafficHistoryPoint> {
        val text = send(request("/metrics/history", mapOf(
            "type" to "traffic",
            "id" to id.toString(),
            "start" to start.toString(),
            "end" to end.toString(),
            "max_points" to maxPoints.toString(),
            "interval" to interval
        )))
        return json.decodeFromString(HistoryResponse.serializer(TrafficHistoryPoint.serializer()), text).data
    }
}
